controllers/sparkcluster_controller.go

```diff
@@ -447,13 +468,51 @@ func (r *SparkClusterReconciler) reconcileStatefulSets(ctx context.Context, sc *
 		log.V(1).Info("updating status", "path", ".status.workerSelector", "value", sc.Status.WorkerSelector)
 	}
 
+	var status v1.JobConditionType
+	masterPod, masterPodFound := r.getMasterPod(ctx, sc)
+	if masterPodFound {
+		if dcv1alpha1.IsPodReady(masterPod) {
+			status = dcv1alpha1.RunningStatus
+		} else {
+			status = dcv1alpha1.StartingStatus
+		}
+	} else {
+		status = dcv1alpha1.PendingStatus
+	}
+	if sc.Status.ClusterStatus != status && !hasDeletionTimestamp(sc) {
+		updateStatus = true
+		sc.Status.ClusterStatus = status
+		if status == dcv1alpha1.RunningStatus {
+			tt := metav1.Now()
+			sc.Status.StartTime = &tt
+		} else {
+			sc.Status.StartTime = nil
+		}
+	}
+
 	if updateStatus {
 		err = r.Status().Update(ctx, sc)
 	}
 
 	return err
 }
 
+func (r *SparkClusterReconciler) getMasterPod(ctx context.Context, sc *dcv1alpha1.SparkCluster) (corev1.Pod, bool) {
+	opts := []client.ListOption{
+		client.InNamespace(sc.Namespace),
+		client.MatchingLabels(spark.MetadataLabelsWithComponent(sc, spark.ComponentMaster)),
+	}
+	pods := &corev1.PodList{}
+	err := r.List(ctx, pods, opts...)
+	if err != nil && !apierrors.IsNotFound(err) {
+		r.getLogger(ctx).V(1).Error(err, "reading master status")
```

Comment:
If we are here, will `Items` be defined below? Or should we just return empty list / false?

Reply:
I am pretty sure that `Items` is initialized an en empty slice. I will test and/or correct.